### ISSUE TITLE
Vote: a new repository for implementation of the Time-Stamp Protocol (TSP)

### ISSUE THREAD

**Issue by yizha1** [notaryproject/.github#58]
I would like to propose the creation of a new repository for implementation of timestamp ([RFC3161](https://www.rfc-editor.org/rfc/rfc3161)).  Here are the details:

Repository name: (Open to suggestions)

`timestamp` 

Description: 

This repository contains an implementation of the Time-Stamp Protocol (TSP) as specified in [RFC3161](https://www.rfc-editor.org/rfc/rfc3161).

@notaryproject/notaryproject-governance-maintainers If you agree with above proposal, please comment `LGTM` or `+1` in this issue.

**Comment by FeynmanZhou** [comment]
LGTM. Linking https://github.com/notaryproject/notation/issues/838 for context

**Comment by shizhMSFT** [comment]
+1

**Comment by priteshbandi** [comment]
+1

**Comment by rgnote** [comment]
IMO, `timestamp` is too generic and doesn't indicate that the package deals with TSA timestamping. 

Below are couple of alternatives I can think of

1. notation-trusted-timestamp
2. notation-tsa-signature

**Comment by shizhMSFT** [comment]
I would like to suggest an alternative name of `timestamp-go` to match other repositories like `notation-go`.

**Comment by shizhMSFT** [comment]
@rgnote TSA stands for Time Stamping Authority but this new repository intends to implement a golang client library. Thus, I don't think it is good to include the term `tsa` in the name.

Another key point is that the repo name is not only just the name of the repository but also the name of the golang module. If we name this repository as `timestamp` or `timestamp-go`, it is easy to have the module name `timestamp` and can be invoked by others as follows.
```go
package main

import "github.com/notaryproject/timestamp-go"

func main() {
	timestamp.Demo()
}
```
If we name the repo name as `trusted-timestamp`, it is difficult to name the package name and harder for users to adopt the library as follows (package alias is required)
```go
package main

import timestamp "github.com/notaryproject/trusted-timestamp"

func main() {
	timestamp.Demo()
}
```
Detailed doc can be found at https://go.dev/wiki/PackagePublishing

**Comment by Two-Hearts** [comment]
+1 for `timestamp-go` based on discussions above.

**Comment by yizha1** [comment]
+1 for `timestamp-go` 

@priteshbandi @FeynmanZhou @toddysm @vaninrao10 @rgnote @gokarnm would you mind voting for the name of the new repo?

**Comment by toddysm** [comment]
We discussed this in the Notary Project community meeting. Naming is still something that we need to resolve. @NiazFK was going to look at the issue and propose a name before Monday's meeting.

**Comment by gokarnm** [comment]
IMO `timestamp` is too generic, a name that includes _timestamp protocol_ or _rfc3161_ would be preferable.

**Comment by yizha1** [comment]
Thanks @gokarnm and @rgnote, as described in the description of this issue

This repository contains implementations of the Time-Stamp Protocol (TSP) as specified in [RFC3161](https://www.rfc-editor.org/rfc/rfc3161) and a subset of pkcs7/CMS (Cryptographic Message Syntax) as specified in ([RFC5652](https://datatracker.ietf.org/doc/html/rfc5652)).

So, it is not purely rfc3161.

**Comment by gokarnm** [comment]
@yizha1 could you share some details about what subset pkcs7/CMS is being implemented. Will that subset be directly available to package consumers through APIs?

**Comment by yizha1** [comment]
@shizhMSFT or @Two-Hearts Could you help to answer the question?

> @yizha1 could you share some details about what subset pkcs7/CMS is being implemented. Will that subset be directly available to package consumers through APIs?

**Comment by Two-Hearts** [comment]
> @yizha1 could you share some details about what subset pkcs7/CMS is being implemented. Will that subset be directly available to package consumers through APIs?

@gokarnm In this new package there's a `BER to DER converter` and an implementation of `signed-data content type of CMS` (ref: https://datatracker.ietf.org/doc/html/rfc5652#section-5). In terms of API level, the current design is mainly related to `rfc 3161 Timestamping`.
For the naming of the new repo, since we have libraries like `notation-go` and `notation-core-go`, `timestamp-go` sounds like a good name to align with other libraries in notary project. But we could discuss more on the naming if you think it's too generic. Do you have any suggestions on your side? If yes, could you list it here? Thanks.

**Comment by gokarnm** [comment]
@Two-Hearts  thanks for the details. If the functionality is scoped to RFC 3161 timestamping, can we name the package as `rfc3161-go`? `timestamp-go` can be misinterpreted as anything to do with generic timestamps, other options are `authentictime-go` , `trustedtime-go` or similar. Do we plan to add any new functionality in this package not related to RFC 3161 in future? If not, I'd vote for `rfc3161-go`.

**Comment by shizhMSFT** [comment]
> @Two-Hearts thanks for the details. If the functionality is scoped to RFC 3161 timestamping, can we name the package as `rfc3161-go`? `timestamp-go` can be misinterpreted as anything to do with generic timestamps, other options are `authentictime-go` , `trustedtime-go` or similar. Do we plan to add any new functionality in this package not related to RFC 3161 in future? If not, I'd vote for `rfc3161-go`.

@gokarnm What if later there is a new RFC doc, which obsolete RFC3161? For example, a COSE-based timestamping protocol.

**Comment by gokarnm** [comment]
I don't think RFC3161 and its support by public CAs will obsolete anytime soon. A COSE based protocol can come up, I think we'd probably implement it in a different package, so we have well defined core timestamping libraries, that consumers can pick, with their own security reviews and advisories. That way we can keep ASN.1 based implementation separate from COSE based. We can implement appropriate abstractions for trusted timestamp if required in the consuming upstream package, but I think it might be too early for that now.

**Comment by shizhMSFT** [comment]
SGTM. Let's put `rfc3161-go` as a candidate name.

**Comment by yizha1** [comment]
Thanks @shizhMSFT @gokarnm , now we have two candidates.

1. `timestamp-go`
2. `rfc3161-go`

Should we start a new round of votes? 

@priteshbandi @FeynmanZhou @toddysm @vaninrao10 @gokarnm @shizhMSFT @Two-Hearts

**Comment by whalelines** [comment]
I think `timestamp-go` is more clear than `rfc3161-go` but I think `timestamp-go` is too generic. The purpose is not a generic library processing timestamps but an implementation of a client for a timestamp authority (TSA). As such, I would recommend something like `tsaclient` or `tsprotocol`.

**Comment by priteshbandi** [comment]
I second David's suggestion of using  `tsa-client-go` as repo name and `tsaclient` as package name.

**Comment by toddysm** [comment]
I agree with @whalelines and @priteshbandi on including `client` in the name and not using the generic `timestamp` or cryptic `rfc3161`. I would suggest `tsprotocol-client-go` or `timestampprotocol-go`.

**Comment by yizha1** [comment]
Now we have more options:

1. `timestamp-go` (original)
2. `rfc3161-go` (by Milind)
3. `tsaclient` (by David)
4. `tsprotocol` (by David)
5. `tsa-client-go` (by Pritesh)
6. `tsprotocol-client-go` (by Toddy)
7. `timestampprotocol-go` (by Toddy)

I vote for option-5 `tsa-client-go`

**Comment by JeyJeyGao** [comment]
`timestamp-go` follows the convention of RFC 3161 package naming (for example, as seen in https://github.com/digitorus/timestamp), as mentioned by @FeynmanZhou. By using this name, it will be easier to search for due to the well-known related repository.

Meanwhile, `tsa-client-go` is a clear name that describes the functionality of the repository.

So, I vote for both `timestamp-go` and `tsa-client-go`.

**Comment by shizhMSFT** [comment]
**TL;DR** I'd vote for `timestamp-go` and a variant `tsprotocol-go` of the candidate ``tsprotocol`.

Here's the reasoning:
1. `timestamp-go`: clear although it is generic
2. `rfc3161-go`: accurate but developers need to deep dive into rfc3161
3. `tsaclient`: `tsaclient` stands for Time-Stamp Authority client. It seems not correct as it is a time-stamping client (i.e. `tsclient-go`).
4. `tsprotocol`: it makes sense as it is a better name than `rfc3161` but I think it should be `tsprotocol-go`.
5. `tsa-client-go`: since golang does not support hyphen `-` in the module name, it is not user friendly as it will be used as `import tsaclient "github.com/notaryproject/tsa-client-go"` (module name alias is always required)
6. `tsprotocol-client-go`: same reason as above
7. `timestampprotocol-go`: the name is too long as a module name.

**Comment by Two-Hearts** [comment]
I vote for `timestamp-go` and `tsaclient-go` (given hyphen is not well supported by golang in module names).

**Comment by FeynmanZhou** [comment]
I vote for `timestamp-go`. 

In the context of Notary Project, existing users and developers might be aware of what's the purpose and usage of this library.

For new users and viewers, a generic but clear project name is good for SEO and easier to memory.

**Comment by vaninrao10** [comment]
repo as "tsaclient-go" and package name "tsaclient".

**Comment by yizha1** [comment]
Based on the discussion during the meeting on Jan 16th, 2024, we vote for the name `tspclient-go` as the repo name, and `tspclient` as the package name.

**Comment by yizha1** [comment]
I will start to create the new repo, set up necessary governance and then update the status later.

**Comment by yizha1** [comment]
Thank you all for voting. Closed this issue as https://github.com/notaryproject/tspclient-go created.
